### DISCUSSION TITLE
Multiple Audiences (possibly via scopes)?

### DISCUSSION THREAD

**Discussion by x2764tech** [simpleidserver/SimpleIdServer#290] [Ideas]
The OpenID Connect spec allows for the `aud` claim to be an array. 

Would it be possible to associate a scope with additional audience(s)? (in our case, so that a Resource Server doesn't have to know valid audiences in advance to validate the token)

Our use case is a public facing API, where we have a dynamic list of Client applications authenticated via OIDC OP. Potentially, the API *could* access this list, but it would be more helpful if the issued tokens could include an additional `aud` if the User consents to the requested scope.

**Comment by simpleidserver** [comment]
Hello,

In our current implementation the `aud` claim only contains the `clientId`.
This implementation is not correct and can be improved :).

As suggested by the RFC 7519 section 4.1.3 (https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3), the audience identifies the recipients that the JWT is intended for.

For example, there are three APIS/CLIENTS with the following configuration :

Id : API1
Scope : api2.read, api3.read

Id : API2

Id : API3

When the API1 is fetching an access token valids on the scope `api2.read` and `api3.read` then the `aud` claim is equals to ["API2", "API3"].
In other Identity solutions, scopes are not added into `aud` claim.
Does-this solution solve your issue ?

If it's the case, then we can make some modification in the release\2.0.11 to support this feature.
We are probably going to introduce the concept of `API Resource`

**Reply by x2764tech** [reply]
That would fix the issue for me.

**Comment by simpleidserver** [comment]
We made some changes in the Solution in order to support the notion of API resources.

1. Install the Prelease-Nuget package into your local OPENID/OAUTH2.0 server.

```
dotnet add package SimpleIdServer.UI.Authenticate.LoginPassword --version 2.0.11-ci-00295 --source https://www.myget.org/F/advance-ict/api/v3/index.json
```

2. Edit the file `OpenIdDefaultConfiguration` and add two clients :

**api1**

api1 has access to the scope `api2.read`.

```
new OpenIdClient
{
    ClientId = "api1",
    ClientSecret = "password",
    ApplicationKind = ApplicationKinds.Service,
    TokenEndPointAuthMethod = "client_secret_post",
    ApplicationType = "web",
    UpdateDateTime = DateTime.UtcNow,
    CreateDateTime = DateTime.UtcNow,
    TokenExpirationTimeInSeconds = 60 * 30,
    RefreshTokenExpirationTimeInSeconds = 60 * 30,
    TokenSignedResponseAlg = "RS256",
    IdTokenSignedResponseAlg = "RS256",
    AllowedScopes = new List<OAuthScope>
    {
        new OAuthScope
        {
            Name = "api2.read"
        }
    },
    GrantTypes = new List<string>
    {
        "client_credentials",
    },
    PreferredTokenProfile = "Bearer",
    ResponseTypes = new List<string>
    {
        "token"
    }
}
```

**api2**

api2 contains one resource named `api2.read`

```
new OpenIdClient
{
    ClientId = "api2",
    ClientSecret = "password",
    ApplicationKind = ApplicationKinds.Service,
    TokenEndPointAuthMethod = "client_secret_post",
    ApplicationType = "web",
    UpdateDateTime = DateTime.UtcNow,
    CreateDateTime = DateTime.UtcNow,
    TokenExpirationTimeInSeconds = 60 * 30,
    RefreshTokenExpirationTimeInSeconds = 60 * 30,
    TokenSignedResponseAlg = "RS256",
    IdTokenSignedResponseAlg = "RS256",
    GrantTypes = new List<string>
    {
        "client_credentials",
    },
    PreferredTokenProfile = "Bearer",
    ResponseTypes = new List<string>
    {
        "token"
    },
    Scopes = new List<ClientScope>
    {
        new ClientScope
        {
            ClientId = "api2",
            Scope = "api2.read"
        }
    }
}
```

3. Execute the following HTTP request to get an access token. The `aud` claim contains the resource `api2`.

You can download and execute the sample solution here.
[ApiResources.zip](https://github.com/simpleidserver/SimpleIdServer/files/9094100/ApiResources.zip)
